PyInstaller/utils/hooks/conda.py

```diff
@@ -0,0 +1,194 @@
+        # Code goes here. e.g.
+        binaries = conda_support.collect_dynamic_libs("numpy")
+        ...
+
+Packages are all referenced by the *distribution name* you use to install it,
+rather than the *package name* you import it with. i.e Use
+``search_distribution("pillow")`` instead of ``search_distribution("PIL")``.
+
+"""
+
+import sys
+from pathlib import Path
+import json
+
+from PyInstaller import compat
+from PyInstaller.log import logger
+
+
+# Conda virtual environments each get their own copy of `conda-meta` so the use
+# of `sys.prefix` instead of `sys.base_prefix`, `sys.real_prefix`  or anything
+# from our `compat` module is intentional.
+CONDA_ROOT = Path(sys.prefix)
+CONDA_META_DIR = CONDA_ROOT / "conda-meta"
+
+
+def search_distribution(name):
+    """
+     Finds and extracts the distribution metadata from the distribution json
+     files in the ``conda-meta`` directory.
+
+    :param name: Distribution name.
+    :type name: str
+    :return: The json-parsed distribution information.
+    :rtype: dict
+    :raises: ModuleNotFoundError: If **distribution** was not installed with
+             conda.
+
+    This function collects a single distribution. But this is generally
+    discouraged due to Conda's often moving distribution contents to
+    base-distributions. e.g. *numpy* itself is empty but depends on
+    *numpy-base*, which contains the python files and depends on *mkl* and
+    *blas* which contain the DLLs. Instead use :meth:`walk_dependency_tree`.
+
+    """
+    # Locate the right json. The name format is:
+    #   [distribution name]-[version][ugly build # hash].json
+
+    # The json name can't be determined exactly so they must be glob searched:
+    for path in CONDA_META_DIR.glob(name + "-*.json"):
+        # But this will include distributions with a prefix of 'name'.
+        # e.g. 'numpy' would also pick up 'numpy-base'.
+        dic = json.loads(path.read_text())
+        if name == dic["name"]:
+            # Read the json to check we've found the right one.
+            # XXX: Maybe we should make this case and '-'/'_' insensitive?
+            return dic
+    raise ModuleNotFoundError("Distribution {} is either not installed or was"
+                              " not installed using Conda.".format(name))
+
+
+def walk_dependency_tree(initial, excludes=None):
+    """
+    Recursively walk the dependencies of a distribution.
+
+    :param initial: Distribution name to collect from.
+    :type initial: str
+    :param excludes: Distributions to exclude, defaults to ``None``.
+    :type excludes: iterable  of str, optional
+    :return: A **distribution_name** to **metadata** mapping where **metadata**
+             is the output of ``search_distribution(distribution_name)``.
+    :rtype: dict
+
+    """
+    if excludes is not None:
+        excludes = set(excludes)
+
+    # Rather than use true recursion, mimic it with a to-do queue.
+    distributions_done = {}
+    names_to_do = {initial}
```

Comment:
Have you considered using a `collections.deque` for this? They're designed for situations like this.

Reply:
You can't efficiently use `in` with a deque so keeping the to-do list duplicate free will become expensive.

Reply:
Actually... it doesn't matter if there are duplicates in the to-do list. The `if _name in distributions_done: continue` will prevent any real work being run more than once. Deque it is then...